neoemf-core/src/main/java/fr/inria/atlanmod/neoemf/core/internal/ContentsList.java

```diff
@@ -0,0 +1,285 @@
+/*
+ * Copyright (c) 2013-2018 Atlanmod, Inria, LS2N, and IMT Nantes.
+ *
+ * All rights reserved. This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License v2.0 which accompanies
+ * this distribution, and is available at https://www.eclipse.org/legal/epl-2.0/
+ */
+
+package fr.inria.atlanmod.neoemf.core.internal;
+
+import fr.inria.atlanmod.commons.LazyReference;
+import fr.inria.atlanmod.commons.primitive.Booleans;
+import fr.inria.atlanmod.neoemf.core.PersistentEObject;
+import fr.inria.atlanmod.neoemf.data.store.Store;
+import fr.inria.atlanmod.neoemf.util.EObjects;
+
+import org.eclipse.emf.ecore.EObject;
+import org.eclipse.emf.ecore.EStructuralFeature;
+import org.eclipse.emf.ecore.impl.EClassImpl;
+import org.eclipse.emf.ecore.util.AbstractSequentialInternalEList;
+import org.eclipse.emf.ecore.util.FeatureMap;
+import org.eclipse.emf.ecore.util.FeatureMapUtil;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.ListIterator;
+
+import javax.annotation.Nonnegative;
+import javax.annotation.Nonnull;
+import javax.annotation.ParametersAreNonnullByDefault;
+import javax.annotation.ParametersAreNullableByDefault;
+
+import static fr.inria.atlanmod.commons.Preconditions.checkPositionIndex;
+import static java.util.Objects.nonNull;
+
+/**
+ * A read-only sequential {@link List} that delegates its operations to the associated {@link Store}.
+ * <p>
+ * Instances of this class are created by {@link PersistentEObject#eContents()} and allows to access the content of a
+ * {@link PersistentEObject} by lazily loading the elements.
+ *
+ * @param <E> the type of elements in this list
+ */
+@ParametersAreNonnullByDefault
+public class ContentsList<E extends EObject> extends AbstractSequentialInternalEList<E> {
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Override the "equals" method in this class. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2160)